packages/smooth_app/lib/pages/prices/generic_infinite_scroll.dart

```diff
@@ -0,0 +1,314 @@
+  Future<void> loadMore(P parameters) async {
+    if (_isLoading || !_hasMoreItems) {
+      return;
+    }
+
+    this.parameters = parameters;
+    _isLoading = true;
+
+    try {
+      final (List<T> newItems, bool hasMore) =
+          await fetchItems(parameters, _currentPage + 1);
+
+      _items.addAll(newItems);
+      _currentPage++;
+      _hasMoreItems = hasMore;
+    } catch (e) {
+      onError?.call(e);
+    } finally {
+      _isLoading = false;
+    }
+  }
+
+  /// Reset the controller to its initial state with optional new initial items
+  void reset({List<T>? newInitialItems}) {
+    _currentPage = initialPage;
+    _items = newInitialItems != null
+        ? List<T>.from(newInitialItems)
+        : List<T>.from(initialItems);
+    _isLoading = false;
+    _hasMoreItems = true;
+  }
+
+  /// Update pagination metadata
+  void updatePaginationInfo({int? newTotalItems, int? newTotalPages}) {
+    totalItems = newTotalItems;
+    totalPages = newTotalPages;
+  }
+}
+
+/// A generic stateful widget for infinite scrolling lists.
+class InfiniteScrollList<T, P> extends StatefulWidget {
+  const InfiniteScrollList({
+    super.key,
+    required this.controller,
+    required this.itemBuilder,
+    required this.loadingBuilder,
+    required this.errorBuilder,
+    required this.emptyBuilder,
+    required this.parameters,
+    this.loadMoreTriggerOffset = 200.0,
+    this.loadingMoreBuilder,
+    this.headerBuilder,
+    this.footerBuilder,
+    this.physics,
+    this.padding,
+    this.shrinkWrap = false,
+    this.onRefresh,
+  });
+
+  /// Controller for managing the infinite scroll behavior
+  final InfiniteScrollController<T, P> controller;
+
+  /// Builder for individual list items
+  final Widget Function(BuildContext context, T item, int index) itemBuilder;
+
+  /// Builder for showing loading state
+  final Widget Function(BuildContext context) loadingBuilder;
+
+  /// Builder for showing error state
+  final Widget Function(BuildContext context, dynamic error) errorBuilder;
+
+  /// Builder for showing empty state
+  final Widget Function(BuildContext context) emptyBuilder;
+
+  /// Optional builder for showing loading more state at the bottom of the list
+  final Widget Function(BuildContext context)? loadingMoreBuilder;
+
+  /// Optional builder for header
+  final Widget Function(BuildContext context)? headerBuilder;
+
+  /// Optional builder for footer
+  final Widget Function(BuildContext context)? footerBuilder;
+
+  /// Parameters for fetching items
+  final P parameters;
+
+  /// Scroll offset from the bottom that triggers loading more items
+  final double loadMoreTriggerOffset;
+
+  /// ScrollPhysics for the ListView
+  final ScrollPhysics? physics;
+
+  /// Padding for the ListView
+  final EdgeInsets? padding;
+
+  /// Whether the ListView should shrink-wrap its contents
+  final bool shrinkWrap;
+
+  /// Callback for pull-to-refresh functionality
+  final Future<void> Function()? onRefresh;
```

Comment:
I don't think we need an `onRefresh`. Please check that and get rid of it if relevant.